### PR TITLE
Pass to hoist bounded allocas.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Earlier attempt to create allocations that are statically bounded in the entry basic block did not cover all cases since at time of bufferization it might not be easy to resolve the static bounds. A new pass is added that hoists such allocations after the fact.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+struct HoistStaticallyBoundAllocationsPass
+    : HoistStaticallyBoundAllocationsBase<HoistStaticallyBoundAllocationsPass> {
+  void runOnOperation() override;
+};
+
+}  // namespace
+
+/// Some uses of a `memref.alloca` can be replaced with a `memref.subview`
+/// easily. Other uses (like a use in a
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:30]: nit: fix comment formatting?

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        })) {
+      allocaOps.push_back(allocaOp);
+      return;
+    }
+  });
+
+  // Hoist the allocas and replace all uses.
+  OpBuilder builder(&getContext());
+  for (auto allocaOp : allocaOps) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE, {
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:62]: I've never seen this one. Wouldn't LLVM_DEBUG be equivalent?

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      llvm::dbgs() << "Alloca Op : ";
+      allocaOp->dump();
+      int numUses = std::distance(allocaOp.getResult().use_begin(),
+                                  allocaOp.getResult().use_end());
+      llvm::dbgs() << " num Uses : " << numUses;
+    });
+    std::optional<Value> replacement =
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:69]: Curious... it's LLVM/MLIR moving from llvm::Optional to std::optional already?

**Review by dcaballe — APPROVED** [review]
Thanks for fixing this! Minor nits.

@matthias-springer for visibility.

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+/// Some uses of a `memref.alloca` can be replaced with a `memref.subview`
+/// easily. Other uses (like a use in a
+/// `scf.yield` or `func.return`) are non-trivial because of compatibility
+/// between types of different SSA values.
+static bool isUseReplacableWithSubview(OpOperand &use) {
+  Operation *user = use.getOwner();
+  if (isa<linalg::LinalgOp, memref::StoreOp, memref::SubViewOp>(user))
+    return true;
+  return false;
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:37]: For now we can just do that.

File: compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/HoistStaticallyBoundAllocations.cpp:69]: Yeah LLVM/MLIR already moved to use `std::optional`. IIUC `llvm::Optional` is now just an alias of `std::optional`

**Pull request merged**